contrib/gin-gonic/gin/gintrace.go

```diff
@@ -10,117 +10,61 @@ import (
-	return MiddlewareTracer(service, tracer.DefaultTracer)
-}
-
-// MiddlewareTracer returns middleware that will trace requests with the given
-// tracer.
-func MiddlewareTracer(service string, t *tracer.Tracer) gin.HandlerFunc {
-	t.SetServiceInfo(service, "gin-gonic", ext.AppTypeWeb)
-	mw := newMiddleware(service, t)
-	return mw.Handle
-}
+	// TODO(gbbr): Handle this when we switch to OpenTracing.
+	t := tracer.DefaultTracer
+	t.SetServiceInfo(service, "gin-gonic/gin", ext.AppTypeWeb)
+	return func(c *gin.Context) {
+		// bail out if tracing isn't enabled
+		if !t.Enabled() {
+			c.Next()
+			return
+		}
 
-// middleware implements gin middleware.
-type middleware struct {
-	service string
-	trc     *tracer.Tracer
-}
+		resource := c.HandlerName()
+		// TODO(x): get the span from the request context
+		span := t.NewRootSpan("http.request", service, resource)
+		defer span.Finish()
 
-func newMiddleware(service string, trc *tracer.Tracer) *middleware {
-	return &middleware{
-		service: service,
-		trc:     trc,
-	}
-}
+		span.Type = ext.HTTPType
+		span.SetMeta(ext.HTTPMethod, c.Request.Method)
+		span.SetMeta(ext.HTTPURL, c.Request.URL.Path)
 
-// Handle is a gin HandlerFunc that will add tracing to the given request.
-func (m *middleware) Handle(c *gin.Context) {
+		// pass the span through the request context
+		c.Set(spanKey, span)
 
-	// bail if not enabled
-	if !m.trc.Enabled() {
+		// serve the request to the next middleware
 		c.Next()
-		return
-	}
-
-	// FIXME[matt] the handler name is a bit unwieldy and uses reflection
-	// under the hood. might be better to tackle this task and do it right
-	// so we can end up with "user/:user/whatever" instead of
-	// "github.com/foobar/blah"
-	//
-	// See here: https://github.com/gin-gonic/gin/issues/649
-	resource := c.HandlerName()
-
-	// Create our span and patch it to the context for downstream.
-	span := m.trc.NewRootSpan("gin.request", m.service, resource)
-	c.Set(key, span)
-
-	// Pass along the request.
-	c.Next()
 
-	// Set http tags.
-	span.SetMeta(ext.HTTPCode, strconv.Itoa(c.Writer.Status()))
-	span.SetMeta(ext.HTTPMethod, c.Request.Method)
-	span.SetMeta(ext.HTTPURL, c.Request.URL.Path)
+		span.SetMeta(ext.HTTPCode, strconv.Itoa(c.Writer.Status()))
 
-	// Set any error information.
-	var err error
-	if len(c.Errors) > 0 {
-		span.SetMeta("gin.errors", c.Errors.String()) // set all errors
-		err = c.Errors[0]                             // but use the first for standard fields
+		if len(c.Errors) > 0 {
+			span.SetMeta("gin.errors", c.Errors.String())
+			span.SetError(c.Errors[0])
+		}
 	}
-	span.FinishWithErr(err)
 }
 
-// Span returns the Span stored in the given Context and true. If it doesn't exist,
-// it will returns (nil, false)
-func Span(c *gin.Context) (*tracer.Span, bool) {
-	if c == nil {
-		return nil, false
-	}
-
-	s, ok := c.Get(key)
+// Span returns the Span stored in the given Context, otherwise nil.
+func SpanFromContext(c *gin.Context) *tracer.Span {
```

Comment:
should we expose this method to our users? in one of our Go examples (using OpenTracing API) we suggested to use a more generic approach: https://github.com/DataDog/dd-opentracing-go/blob/master/main.go#L38

I think would be great avoiding too many `SpanSomethingContext` methods, since I would prefer only one way (and so one API) to extract a `Span` from a `Context`. Unless, of course, we hit limitations in some frameworks. Do you think this is legit?